jep/0001/README.adoc

```diff
@@ -0,0 +1,306 @@
+which produces a `ExternalLoggingEventWriter` class instance.
+** This method automatically injects the mandatory metadata for events
+** An abstract `_createWriter()` method is offered for implementation by downstream implementations
+* Being compared to `LoggingMethod`,
+the implementation is designed to always perform logging
+from the agent side
+
+ExternalLoggingEventWriter class:
+
+* The class is `Serializable`.
+It will be sent to the agent side.
+* The class offers the following abstract methods:
+** `void writeEvent(Event event) throws IOException`
+- writing event to the remote storage
+* The class also stores metadata, which may be injected into the events
+** The class stores a Map of Serializable metadata entries
+** The class offers API, which allow setting the metadata.
+   This API will be used by `ExternalLoggingMethod` implementations and other logic
+   to provide additional metadata if required
+
+=== ExternalLogBrowserFactory and classes
+
+This factory just produces instances of `ExternalLogBrowser`.
+This class offers an abstraction layer for external log browsing.
+
+`ExternalLogBrowser` will also provide an abstraction layer for
+eventual consistency management.
+This layer will be determined during reference implementation polishing.
+
+=== Pipeline Bridge extensions
+
+The plugin should also implement Pipeline `LogStorage` and `LogStorageFactory` extension points
+so that it transparently supports Pipeline with existing API.
+
+Pipeline Storage JEP is documented in JEP-TODO.
+
+=== JSON-based external-logging layer
+
+The API Plugin should offer a convenience layer in order
+to support a number of most common logging providers like
+Logstash, Fluentd, Elasticsearch, AWS CloudWatch, etc.
+
+This layer should provide the following features:
+
+* Utility classes for reading and writing JSON Events,
+including parsing/writing `ConsoleNote` objects
+* Base classes for constructing JSON queries for fetching data
+
+Other convenience layers will be defined during prototyping.
+
+=== Configuration
+
+External Logging API plugin should be fully configurable via
+WebUI and _Configuration-as-Code Plugin_.
+It includes:
+
+* Selection of `LoggingMethod` and `LogStorage` factories
+* Configuration of built-in External Logging and Log Browser factories
+* Any other configuration options
+
+=== Reference implementation
+
+As a reference implementation of the _External Logging API_ plugin,
+a new _External Logging for Elasticsearch Plugin_ will be implemented.
+Other implementations may be also created.
+
+== Reasoning
+
+=== Why separate plugin?
+
+One may say that all bits in the current design could be implemented
+as a part of the Jenkins core.
+It is true, but detaching of the plugin is intentional:
+
+* The plugin will have a separate release cycle so that changes in it
+  can be delivered and backported independently from the core's release cycle
+* The approach allows keeping the patches on the core's side minimal
```

Comment:
Is it really minimal? Minimal core APIs would be some extension point just allowing a plugin to override the behavior of certain critical methods.

Reply:
It's pretty much what JEP-207 does. `LoggingMethod` + `LogBrowser` + `LoggingMethodLocator`